<launch>
  <!-- General arguments -->
  <arg name="use_batch_mode"
       default="true"
       doc="Whether to process the rosbag in batch mode or replay the rosbag."/>
  <arg name="param_file"
       default="$(find wavemap_3d_ros)/config/ouster_os0.yaml"
       doc="Name of the file from which to read wavemap's server params."/>
  <arg name="debug_mode"
       default="false"
       doc="Run the mapper in a debugger (gdb)."/>
  <arg name="show_rviz"
       default="true"
       doc="Automatically launch Rviz."/>
  <arg name="record_rosbag" default="false"/>

  <!-- Wavemap params -->
  <arg name="data_structure_type"
       default="hashed_blocks"/>
  <arg name="measurement_model_type"
       default="single_ray_integrator"/>

  <!-- Dataset params -->
  <arg name="pointcloud_topic"
       default="/os_cloud_node/points"/>
  <arg name="rosbag_path"
       default="
../data/newer_college/cloister/2021-12-02-10-15-59_0-cloister.bag
../data/newer_college/cloister/2021-12-02-10-19-05_1-cloister.bag"
       doc="Replay the rosbag at the specified path."/>
  <!-- NOTE: Relative paths will be resolved w.r.t. ROS_HOME (default ~/.ros) -->
  <node pkg="tf" type="static_transform_publisher"
        name="static_tf_cam_left_os1_lidar"
        args="1e-3 0 90.683e-3 0 0 0 1 base os_sensor 100"/>
  <node pkg="wavemap_3d_ros" type="csv_to_tf.py" name="csv_to_tf_pub"
        output="screen">
    <param name="ground_truth_pose_file_path"
           value="../data/newer_college/cloister/gt-nc-cloister.csv"/>
    <param name="sensor_frame"
           value="base"/>
  </node>


  <!-- Process the rosbag with wavemap -->
  <group if="$(arg use_batch_mode)">
    <!-- In batch mode (process the rosbag as fast as possible) -->
    <include file="$(find wavemap_3d_ros)/launch/rosbag_processor.launch"
             pass_all_args="true"/>
  </group>
  <group unless="$(arg use_batch_mode)">
    <!-- In normal mode (replay the rosbag) -->
    <include file="$(find wavemap_3d_ros)/launch/wavemap_server.launch"
             pass_all_args="true"/>
  </group>

  <!-- Record rosbag -->
  <group if="$(arg record_rosbag)">
    <!-- Relative paths will be resolved w.r.t. ROS_HOME (default ~/.ros) -->
    <arg name="bag_save_path" default="../catkin_ws/bags/evaluations"/>
    <arg name="bag_prefix"
         default="simple_building_manual-$(arg data_structure_type)-$(arg measurement_model_type)"/>
    <node pkg="rosbag" type="record" name="rosbag_record" output="screen"
          args="/wavemap_3d/map_evaluation_summary /wavemap_3d/performance_stats /wavemap_3d/occupancy_grid /wavemap_3d/occupancy_grid_error --output-prefix $(arg bag_save_path)/$(arg bag_prefix)"/>
  </group>
</launch>
